src/Storages/StorageMerge.cpp

```diff
@@ -146,8 +146,43 @@ bool StorageMerge::canMoveConditionsToPrewhere() const
     /// If new table that matches regexp for current storage and doesn't support PREWHERE
     /// will appear after this check and before calling "read" method, the optimized query may fail.
     /// Since it's quite rare case, we just ignore this possibility.
+    const auto & table_doesnt_support_prewhere = getFirstTable([](const auto & table) { return !table->canMoveConditionsToPrewhere(); });
+    bool can_move = (table_doesnt_support_prewhere == nullptr);
 
-    return getFirstTable([](const auto & table) { return !table->canMoveConditionsToPrewhere(); }) == nullptr;
+    if (!can_move)
+        return false;
+
+    if (!getInMemoryMetadataPtr())
```

Comment:
Hi @vdimir. Is this at all possible? `setInMemoryMetadata` in `StorageMerge` should be enough, no?

Reply:
I've removed it here - #46454, it also addresses few other issues with PREWHERE for Merge, and this patch makes it possible to use PREWHERE for columns which types is the same, instead of globally disable PREWHERE